src/box/memtx_tree.cc

```diff
@@ -1377,16 +1338,16 @@ memtx_tree_func_index_replace(struct index *base, struct tuple *old_tuple,
 		 * replaced entries.
 		 */
 		rlist_foreach_entry(undo, &old_keys, link) {
-			func_index_key_free((const char *)undo->key.hint);
+			tuple_unref((struct tuple *)undo->key.hint);
 		}
 	}
 	if (old_tuple != NULL) {
 		if (key_list_iterator_create(&it, old_tuple, index_def, false,
-					     func_index_key_dummy_alloc) != 0)
+					     memtx->func_key_format) != 0)
 			goto end;
 		struct memtx_tree_data<true> data, deleted_data;
 		data.tuple = old_tuple;
-		const char *key;
+		struct tuple *key;
 		while (key_list_iterator_next(&it, &key) == 0 && key != NULL) {
```

Comment:
So we have to allocate tuple for deletion from space.. Deadlock is possible.

Reply:
Right. Unfortunately, I can't use `memtx_tree_find` here, as you proposed f2f yesterday, because I'd need a full key for that (including primary key parts), i.e. another allocation. I fixed the potential deadlock by using `tuple_format_runtime` instead of `memtx->func_key_format` here. It's okay, because we need these keys only to delete all tuple references from the index tree (we aren't going to store them in the index tree) so we don't need to delay freeing them in presence of a read view.

<details><summary>diff</summary>

```diff
diff --git a/src/box/memtx_tree.cc b/src/box/memtx_tree.cc
index ba0b1d368880..1ec614f6cf32 100644
--- a/src/box/memtx_tree.cc
+++ b/src/box/memtx_tree.cc
@@ -1342,8 +1342,13 @@ memtx_tree_func_index_replace(struct index *base, struct tuple *old_tuple,
 		}
 	}
 	if (old_tuple != NULL) {
+		/*
+		 * Use the runtime format to avoid OOM while deleting a tuple
+		 * from a space. It's okay, because we are not going to store
+		 * the keys in the index.
+		 */
 		if (key_list_iterator_create(&it, old_tuple, index_def, false,
-					     memtx->func_key_format) != 0)
+					     tuple_format_runtime) != 0)
 			goto end;
 		struct memtx_tree_data<true> data, deleted_data;
 		data.tuple = old_tuple;
```

</details>